Design/indexing_rollout_notes.md

```diff
@@ -0,0 +1,209 @@
+
+
+## General goals
+
+- Quite a few algorithms take an input workspace, extract information about spectrum numbers and detector IDs, and initialize a new workspace based on those data. Example include `SumSpectra`, `DiffractionFocussing`, and `ExtractSpectra`.
+  One goal of the indexing library is to remove explicit uses of spectrum numbers and detector IDs in as much code as possible.
+  Since we need to refactor algorithms similar to those mentioned above, it is probably a good idea to do it properly right away.
+  That is, we should *not* just refactor the manipulation of spectrum numbers and detector IDs via `IndexInfo`.
+  Instead, we should provide functionality for grouping and extraction of spectra to create a new `IndexInfo`, without the need for explicit handling of spectrum numbers and detector IDs.
+
+  As an example, consider
+
+  ```cpp
+  namespace Indexing {
+  IndexInfo extract(const IndexInfo &, const SpectrumIndexSet &);
+  IndexInfo group(const IndexInfo &, const std::vector<SpectrumIndexSet> &);
+  }
+  ```
+
+- Should we set default detector IDs?
+  This is was workspaces are doing currently, but it does not really make sense.
+  Probably we should find a better way if setting sensible detector IDs once an instrument has been set.
+  This is something that potentially needs to be decided together with other decisions regarding Instrument-2.0.
+
+
+## NearestNeighbours
+
+- Uses `specnum_t` as map key.
+- Should simply use spectrum index?
+- Would that make trouble with MPI? Need to have a detailed look.
+- See issue [#17372](https://github.com/mantidproject/mantid/issues/17372).
+
+
+## Logging and exception messages
+
+- Frequently spectrum numbers are included in log messages or exceptions.
+
+- Probably the easiest (overall) solution is to add means of getting this information automatically, e.g., with a helper function
+
+  ```cpp
+  namespace Logging {
+  std::string spectrumNumber(const MatrixWorkspace &workspace, size_t index);
+  }
+  ```
+
+## File- and UI output
+
+Similar to logging, there are also several algorithms that write, e.g., detector IDs to a file.
+Maybe a similar or even the same helper function as suggested above in the logging section could be used.
+
+
+## Obtaining detector information
+
+Frequently, detector IDs are used to obtain detector pointers or other detector-related information from the instrument.
+The IDs are not actually required here.
+Refactoring would be significantly easier (and would avoid the need to revisit and incrementally refactor) if such information was available via, e.g., `MatrixWorkspace::SpectrumInfo`.
+
+Commonly used functionality includes:
+
+```cpp
+class SpectrumInfo {
+  bool isMasked(const size_t index) const;
+  bool isMonitor(const size_t index) const;
+  double l2(const size_t index) const;
+  Detector_sptr uniqueDetector(const size_t index) const; // Throws if no or more than 1 detector
```

Comment:
@OwenArnold In what I have implemented so far in https://github.com/mantidproject/mantid/pull/17394 I am actually not providing access to detectors. I think it would be used in some places, but I am not sure that this belongs into `SpectrumInfo`. I think the key question for this is:
*Does Instrument-2.0 provide/need detector groups?*
Currently detector groups are used for things like defining which detectors belong to a spectrum, and for passing it around as a single object, for reading parameters (which may get averaged for all detectors in the group). Is there anything in this list that Instrument-2.0 cannot solve better in another way?
- `SpectrumInfo` covers many of the cases. For example, we can get L2 for a spectrum, which is defined as the average of all associated detectors. This does not require groups.
- `ParameterMap` access? This is still unclear.
- Others?

If we do not have detector groups `SpectrumInfo` cannot and should not be able to return detector pointers or references. I might be able to return containers, with zero or more entries. One can argue that a detector group is just a container, but actually the composite pattern used in Instrument-1.0 is maybe not ideal here. For example, `IDetector::getID()` returns the detector ID of the first detector in the group. This seems wrong. Maybe forcing this into a common schema is just not the right thing here?